### PR TITLE
Bugfix/1316 wrong conversion pipette resistance

### LINKED ISSUE #1315: USER_Pipette in Bath pipette resistance incorrect units
MIES version: Release_2.4_20211216-327-g616735c6

The "USER_Pipette in Bath pipette resistance" appears to be in units of milliohms rather than ohms. I tested this with a model cell in the "Bath" configuration which simulates a 10 MOhm pipette in the bath solution and when I ran the "EXTPINBATH180424_DA_0" sweep I noticed that the value being returned for "USER_Pipette in Bath pipette resistance" was about 10,000,000,000 Ohms, which would be 10 GOhms. While I still had the model cell in the same "Bath" configuration I also ran the EXTPCllATT180424_DA_0 sweep which uses "USER_Seal evaluation seal resistance max" and that value was returning a correct value of about 10,000,000 Ohms or about 10 MOhms

This image shows the published output of both the "USER_Seal evaluation seal resistance max" (top circle) and the "USER_Pipette in Bath pipette resistance" (bottom circle) and although they should both be around 10,000,000 Ohms (10 MOhms), the "USER_Pipette in Bath pipette resistance" is 10,000,000,000 Ohms (10 GOhms), and they should both be returning roughly the same value.
labnotebook_inbath_resistances_igorrouter_LI

This image shows the lab notebook entry for the "USER_Pipette in Bath pipette resistance". I placed a cursor over the last data point (top circle) and the resistance value is 1.0537 x 10 ^10 (10.5 billion Ohms, or about 10.5 GOhms)
labnotebook_inbath_resistances_LI

I also have the pxp file that I was using for testing that I will upload to the FTP server (2022_03_22_pipette_resistance_units.pxp)

Please let me know if you have any questions or need any additional information about the issue.

### PR DESCRIPTION
**Pull request by t-b**
- [x] Add percentage conversion constants
- [x] Add linting check with NOLINT support
- [x] Finish

Close #1315.

### CODE REVIEW

**Comment by timjarsky** [comment]
@t-b converting from MIES units to standard units is something we do reasonably often? If yes, should we use constants? e.g., MegaOhmToOhm, msToSec, pAToAmp, mVToV

**Comment by t-b** [comment]
> @t-b converting from MIES units to standard units is something we do reasonably often? If yes, should we use constants? e.g., MegaOhmToOhm, msToSec, pAToAmp, mVToV

Yes, we should do that. Although I would leave away the units and only use the scales, e.g. MEGA_TO_ONE, MEGA_TO_MILLI, etc.

**Comment by t-b** [comment]
@MichaelHuth Please have a thorough look.

File: Packages/MIES/MIES_Constants.ipf [hunk 1]
```diff
@@ -1718,3 +1718,478 @@ Constant DECIMATION_MINMAX = 0x1
...
+/// .. code-block:: igorpro
+///
+///    // pA -> A
+///    value_A = value_pA * PICO_TO_ONE
+///
+///    // V -> mV
+///    value_mV = value_V * ONE_TO_MILLI
+/// \endrst
+///
+/// Generated code from GenerateMultiplierConstants().
```
Comment (timjarsky) [inline Packages/MIES/MIES_Constants.ipf:1773]: it's cool that code was used to generate the constant list.

Comment (timjarsky) [reply Packages/MIES/MIES_Constants.ipf:1773]: @t-b was code also used to find and replace using the constants or is it necessary to hand check each replacement?

Comment (t-b) [reply Packages/MIES/MIES_Constants.ipf:1773]: This can't be done automatically. I did every replacement by hand, this is also the reason why it is grouped in the commits by constants.

For example the factor 1e-3 is either MILLI_TO_ONE or MICRO_TO_MILLI, it depends on the context which one is correct.

File: Packages/Testing-MIES/UTF_SweepFormula.ipf [hunk 2]
```diff
@@ -874,12 +874,12 @@ static Function TestAPFrequency()
 
 	// works with 2D data and instantaneous
 	WAVE output = SF_FormulaExecutor(DirectToFormulaParser("apfrequency(setscale([[10, 5], [20, 40], [10, 5], [20, 30]], x, 0, 5, ms), 1, 15)"))
-	Make/FREE/D output_ref = {100, 94.59459459459458}
+	Make/FREE/D output_ref = {100, 94.59459459459457}
```
Comment (MichaelHuth) [inline Packages/Testing-MIES/UTF_SweepFormula.ipf:877]: Why was this change required?
Is it because a change from /1000 to *1E-3?

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
I think there are a some locations missing, where the new constants should be used:

Function AI_GetMCCScale for MIES to MCC units
https://github.com/AllenInstitute/IPNWB/blob/f3493461480eaafa97939940693ed693860048a6/IPNWB_Debugging.ipf#L241

https://github.com/AllenInstitute/MIES/blob/2c2249f81006995c91648bfd39d4deca0e794966/Packages/MIES/MIES_AmplifierInteraction.ipf#L979
and 2 lines further.

https://github.com/AllenInstitute/MIES/blob/2c2249f81006995c91648bfd39d4deca0e794966/Packages/MIES/MIES_AnalysisFunctions_PatchSeq.ipf#L534

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_DAC-Hardware.ipf#L2038

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_DAC-Hardware.ipf#L2138

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_DataConfigurator.ipf#L2059

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_Epochs.ipf#L203

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_Epochs.ipf#L548
and the following line.

https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_PulseAveraging.ipf#L1335
and all other sprintf execTime lines in PA.

There are some conversions in sweepformula from s to ms, like (search for 1E3):
https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_SweepFormula.ipf#L1032

I am not 100% sure, but these factors for NWB seem to be also unit conversion factors:
https://github.com/AllenInstitute/MIES/blob/234dd29a2184c3e017f73ed6c0aff28563351aa2/Packages/MIES/MIES_NeuroDataWithoutBorders.ipf#L1782
and the following lines.

Searching for "1000", "1000000" and even "1000000000" shows also more locations where units are converted.

File: Packages/Testing-MIES/UTF_SweepFormula.ipf (hunk 2 above)
Comment (t-b) [reply Packages/Testing-MIES/UTF_SweepFormula.ipf:877]: Yes, exactly.

**Comment by t-b** [comment]
> I think there are a some locations missing, where the new constants should be used:

Thanks for the thorough review. Looks like my grep invocation was case sensitive and not finding 1EX. I'll take care of those as well.

**Comment by t-b** [comment]
I'm onto it. My current search expression:

```
git grep -i -I -P -e '(?<!tol=)(?<!tol =)(?<!tol = )(?<!tol= )(\*|/)[[:space:]]*(10+|1e(-|\+)?[[:digit:]]+)\b' --and --not -e "^(Str)?Constant" --and --not -e "(REQUIRE|CHECK)_(EQUAL|CLOSE)_(WAVES|VAR)\(" --and --not -e "^[[:space:]]*//" '*.ipf'`
```

**Review by timjarsky — DISMISSED** [review]

**Comment by t-b** [comment]
@MichaelHuth I've converted a lot of more places, add linting, and also moved the constants to their own file. I did not touch anything in Packages/IPNWB as that is a separate git repo. I also did not squash the commits to ease review. The old commits are unchanged.

**Comment by timjarsky** [comment]
@t-b I've finished checking all the commits. I have an inline comment for b7929c63 but other than that, I can sign off on this wonderfully organized PR.

- [x] fed997b5 (ParseUnit: Prefer constants and move documentation, 2022-03-29) 
- [x] a278e949 (Add constants for SI prefix conversion, 2022-03-29)
- [ ] b7929c63 (Prefer ONE_TO_PICO over * 1e12 and  / 1e-12, 2022-03-29)
- [x] bd9c8afb (Prefer ONE_TO_MEGA over * 1e-6, 2022-03-29)
- [x] 366c3aa8 (Prefer PICO_TO_ONE over * 1e-12, 2022-03-29)
- [x] 09a4d72f (Prefer MILLI_TO_ONE over * 1e-3 or / 1e3 or / 1000, 2022-03-29)
- [x] e95f21bb (Prefer ONE_TO_MILLI over * 1e3, 2022-03-29)
- [x] e55d070f (Prefer KILO_TO_ONE over 1e3, 2022-03-29)
- [x] fec5ba9f (Prefer GIGA_TO_ONE over 1e9, 2022-03-29)
- [x] 5b9de2f5 (Prefer MEGA_TO_ONE over 1e6, 2022-03-29)
- [x] f6eaa823 (Prefer GIGA_TO_MEGA over 1e3, 2022-03-29)
- [x] b7f9ae78 (Prefer ONE_TO_MICRO over 1e6, 2022-03-29)
- [x] e70dff98 (Prefer MICRO_TO_MILLI over 1e-3, 2022-03-29)
- [x] 3efdb67a (Prefer MICRO_TO_ONE over / 1e6, 2022-03-29)
- [x] adda42dc (Prefer MILLI_TO_MICRO over 1e3, 2022-03-29)
- [x] 429c9a0a (PSQ_PipetteInBath: Store resistance values with correct decimal multiplier, 2022-03-29)
- [x] 2cbeb18d (Add tests for ConvertRateToSamplingInterval and ConvertSamplingIntervalToRate, 2022-03-29)
- [x] e26a0238 (ConvertRateToSamplingInterval/ConvertSamplingIntervalToRate: Prefer constants, 2022-03-29)
- [x] 5b54ea71 (AD_PlotBounds: Fix baselineVoltage conversion, 2022-03-29)
- [x] 4e524c39 (tools/check-code.sh: Search for literal decimal multiplier, 2022-04-05) - assumed that all constants are correct
- [x] f47bb943 (squash Add constants for SI prefix conversion, 2022-04-05)
- [x] 102b6758 (UTF_SweepFormula.ipf: Adjust tests to not trigger decimal multiplier linting, 2022-04-06)
- [x] 96ec1df8 (Tests: Use NOLINT to ignore decimal multiplier linting errors, 2022-04-06)
- [x] 93f2bf3f (WAVEBUILDER_MIN_SAMPINT: Add comment with unit, 2022-04-06)

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
See comments for some of the // NOLINTS.  The rest looks good.

File: Packages/MIES/MIES_AmplifierInteraction.ipf [hunk 3]
```diff
@@ -719,7 +719,7 @@ static Function AI_MIESAutoPipetteOffset(device, headStage)
 
 	ASSERT(clampMode == V_CLAMP_MODE || clampMode == I_CLAMP_MODE, "Headstage must be in VC/IC mode to use this function")
 	//calculate delta current to reach zero
-	vdelta = (TPResults[%BaselineSteadyState][headstage] * TPResults[%ResistanceSteadyState][headstage]) / 1000 // set to mV
+	vdelta = (TPResults[%BaselineSteadyState][headstage] * TPResults[%ResistanceSteadyState][headstage]) / 1000 // set to mV // NOLINT
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_AmplifierInteraction.ipf:722]: Yes the formula is U = R * I, but the units are a bit more involved, see 4ac8623e (squash NOLINT, 2022-04-13).

File: Packages/MIES/MIES_MiesUtilities.ipf [hunk 4]
```diff
@@ -7603,7 +7603,7 @@ static Function GetSamplingIntervalFromLBN(WAVE numericalValues, variable sweepN
 	samplingInterval = GetLastSettingIndep(numericalValues, sweepNo, "Sampling interval", DATA_ACQUISITION_MODE)
 
 	// convert to integer as that is stored in the config wave
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_MiesUtilities.ipf:7605]: I've created https://github.com/AllenInstitute/MIES/issues/1333.

File: Packages/MIES/MIES_TestPulse.ipf [hunk 5]
```diff
@@ -991,11 +991,11 @@ threadsafe Function/DF TP_TSAnalysis(dfrInp)
 #endif
 
 	if(clampMode == I_CLAMP_MODE)
-		outData[1] = (avgTPSS - avgBaselineSS) / clampAmp * 1000
-		outData[2] = (avgInst - avgBaselineSS) / clampAmp * 1000
+		outData[1] = (avgTPSS - avgBaselineSS) / clampAmp * 1000 // NOLINT
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_TestPulse.ipf:994]: It ain't that easy, but you beat me to it ;) see 26b84517 (TP_TSAnalysis: Use conversions constants instead of NOLINT, 2022-04-13).

**Comment by t-b** [comment]
The commits

26b84517 (TP_TSAnalysis: Use conversions constants instead of NOLINT, 2022-04-13)
4ac8623e (squash NOLINT, 2022-04-13)

are new.

Once you guys are okay, I'll squash the mess and verify that the result is still the same.

**Review by MichaelHuth — DISMISSED** [review]
ok

**Comment by t-b** [comment]
Test failures look like rounding issues will fix once review is done.

```
Collapse | Failed | SweepFormulaHardware#SF_TPTest:ITC                                 SweepFormulaHardware#SF_TPTest:ITC in UTF_SweepFormulaHardware.ipf (8)                                 History | 14 secs
FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x0870498190 Wave2: 0x08704971b0 Dimensions  \|Labels       \|Value      \| ------------\|-------------\|-----------\| [0][0][1][0]\|;sweep0;DA1;;\|1000.000000\|             \|;sweep0;DA1;;\|1000.000000\| [0][1][1][0]\|;sweep1;DA1;;\|1000.000000\|             \|;sweep1;DA1;;\|1000.000000\| [0][2][1][0]\|;sweep2;DA1;;\|1000.000000\|             \|;sweep2;DA1;;\|1000.000000\| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 227 (280), procedure "UTF_SweepFormulaHardware.ipf" FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x081775fc50 Wave2: 0x08704971b0 Dimensions  \|Labels       \|Value      \| ------------\|-------------\|-----------\| [0][0][1][0]\|;sweep0;DA1;;\|1000.000000\|             \|;sweep0;DA1;;\|1000.000000\| [0][1][1][0]\|;sweep1;DA1;;\|1000.000000\|             \|;sweep1;DA1;;\|1000.000000\| [0][2][1][0]\|;sweep2;DA1;;\|1000.000000\|             \|;sweep2;DA1;;\|1000.000000\| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 223 (280), procedure "UTF_SweepFormulaHardware.ipf" FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x081775fb30 Wave2: 0x08704971b0 Dimensions  \|Labels       \|Value      \| ------------\|-------------\|-----------\| [0][0][1][0]\|;sweep0;DA1;;\|1000.000000\|             \|;sweep0;DA1;;\|1000.000000\| [0][1][1][0]\|;sweep1;DA1;;\|1000.000000\|             \|;sweep1;DA1;;\|1000.000000\| [0][2][1][0]\|;sweep2;DA1;;\|1000.000000\|             \|;sweep2;DA1;;\|1000.000000\| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 219 (280), procedure "UTF_SweepFormulaHardware.ipf" FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x087b5594f0 Wave2: 0x08704971b0 Dimensions  \|Labels       \|Value      \| ------------\|-------------\|-----------\| [0][0][1][0]\|;sweep0;DA1;;\|1000.000000\|             \|;sweep0;DA1;;\|1000.000000\| [0][1][1][0]\|

<!--EndFragment-->
</body>
</html>Status 	Test 	Duration
Collapse 	Failed 	SweepFormulaHardware#SF_TPTest:ITC [SweepFormulaHardware#SF_TPTest:ITC in UTF_SweepFormulaHardware.ipf (8)](http://bamboo.corp.alleninstitute.org/browse/MIES-PLAYIN609-TWIO-8/test/case/178867801) [History](http://bamboo.corp.alleninstitute.org/browse/MIES-PLAYIN609-TWIO/test/case/178867801)
	14 secs

FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x0870498190 Wave2: 0x08704971b0 Dimensions  |Labels       |Value      | ------------|-------------|-----------| [0][0][1][0]|;sweep0;DA1;;|1000.000000|             |;sweep0;DA1;;|1000.000000| [0][1][1][0]|;sweep1;DA1;;|1000.000000|             |;sweep1;DA1;;|1000.000000| [0][2][1][0]|;sweep2;DA1;;|1000.000000|             |;sweep2;DA1;;|1000.000000| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 227 (280), procedure "UTF_SweepFormulaHardware.ipf"
FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x081775fc50 Wave2: 0x08704971b0 Dimensions  |Labels       |Value      | ------------|-------------|-----------| [0][0][1][0]|;sweep0;DA1;;|1000.000000|             |;sweep0;DA1;;|1000.000000| [0][1][1][0]|;sweep1;DA1;;|1000.000000|             |;sweep1;DA1;;|1000.000000| [0][2][1][0]|;sweep2;DA1;;|1000.000000|             |;sweep2;DA1;;|1000.000000| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 223 (280), procedure "UTF_SweepFormulaHardware.ipf"
FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x081775fb30 Wave2: 0x08704971b0 Dimensions  |Labels       |Value      | ------------|-------------|-----------| [0][0][1][0]|;sweep0;DA1;;|1000.000000|             |;sweep0;DA1;;|1000.000000| [0][1][1][0]|;sweep1;DA1;;|1000.000000|             |;sweep1;DA1;;|1000.000000| [0][2][1][0]|;sweep2;DA1;;|1000.000000|             |;sweep2;DA1;;|1000.000000| : is false. Assertion "CHECK_EQUAL_WAVES(tpResult, wRef)" failed in line 219 (280), procedure "UTF_SweepFormulaHardware.ipf"
FAIL: Assuming equality using mode WAVE_DATA for waves _free_ and _free_; detailed: Waves difference: Wave1: 0x087b5594f0 Wave2: 0x08704971b0 Dimensions  |Labels       |Value      | ------------|-------------|-----------| [0][0][1][0]|;sweep0;DA1;;|1000.000000|             |;sweep0;DA1;;|1000.000000| [0][1][1][0]|
```

**Review by timjarsky — DISMISSED** [review]

**Comment by t-b** [comment]
@timjarsky @MichaelHuth Thanks both. Sorry you had to go through that. I'll fix the failing tests tomorrow, polish the commits and ensure that I don't break anything.

**Comment by t-b** [comment]
@timjarsky I've checked that my commit cleaning up did not change anything. I also fixed the failing test.

**Review by timjarsky — APPROVED** [review]
